File: design-pattern/chain-of-responsibility-pattern/crp.ts
abstract class CashDispenser {
    protected cashdispenser: CashDispenser | null;

    constructor() {
        this.cashdispenser = null;
    }

    setHandler(handler: CashDispenser): void {
        this.cashdispenser = handler;
    }

    abstract moneyDispense(amount: number): void;
}


class ThounsandDispenser extends CashDispenser {
    private number_of_notes: number;

    constructor(numnotes: number) {
        super();
        this.number_of_notes = numnotes;
    }

    moneyDispense(amount: number): void {
        let notes_need: number = Math.floor(amount / 1000);

        if (notes_need > this.number_of_notes) {
            notes_need = this.number_of_notes;
        }

        this.number_of_notes -= notes_need;

        if (notes_need > 0) {
            console.log(`dispensing ${notes_need} of $1000 from atm`);
        }

        let remaining_amount = amount - (notes_need * 1000);

        if (remaining_amount > 0) {
            if (this.cashdispenser) {
                this.cashdispenser.moneyDispense(remaining_amount);
            } else {
                console.log(`notes not available to proceed : amount $${remaining_amount}`);
            }
        }
    }
}


class HundredDispenser extends CashDispenser {
    private number_of_notes: number;

    constructor(numnotes: number) {
        super();
        this.number_of_notes = numnotes;
    }

    moneyDispense(amount: number): void {
        let notes_need: number = Math.floor(amount / 100);

        if (notes_need > this.number_of_notes) {
            notes_need = this.number_of_notes;
        }

        this.number_of_notes -= notes_need;

        if (notes_need > 0) {
            console.log(`dispensing ${notes_need} of $100 from atm`);
        }

        let remaining_amount = amount - (notes_need * 100);

        if (remaining_amount > 0) {
            if (this.cashdispenser) {
                this.cashdispenser.moneyDispense(remaining_amount);
            } else {
                console.log(`notes not available to proceed : amount $${remaining_amount}`);
            }
        }
    }
}


class FiftyDispenser extends CashDispenser {
    private number_of_notes: number;

    constructor(numnotes: number) {
        super();
        this.number_of_notes = numnotes;
    }

    moneyDispense(amount: number): void {
        let notes_need: number = Math.floor(amount / 50);

        if (notes_need > this.number_of_notes) {
            notes_need = this.number_of_notes;
        }

        this.number_of_notes -= notes_need;

        if (notes_need > 0) {
            console.log(`dispensing ${notes_need} of $50 from atm`);
        }

        let remaining_amount = amount - (notes_need * 50);

        if (remaining_amount > 0) {
            console.log(`notes not available to proceed : amount $${remaining_amount}`);
        }
    }
}


// setup
const hundredHandler = new HundredDispenser(10);
const thoundsandHandler = new ThounsandDispenser(15);
const fiftyHandler = new FiftyDispenser(20);

thoundsandHandler.setHandler(hundredHandler);
hundredHandler.setHandler(fiftyHandler);

// test
const moneyTodispense = 2500;
thoundsandHandler.moneyDispense(moneyTodispense);

File: design-pattern/chain-of-responsibility-pattern/crp.cpp
#include <iostream>
using namespace std;

class MoneyDispenser
{
protected:
    MoneyDispenser *next;

public:
    MoneyDispenser() : next(nullptr) {}

    void setHandler(MoneyDispenser *m)
    {
        next = m;
    }

    virtual void cashDispenser(int amount) = 0;
    virtual ~MoneyDispenser() {}
};

class ThousandHandler : public MoneyDispenser
{
    int numnotes;

public:
    ThousandHandler(int notes) : numnotes(notes) {}

    void cashDispenser(int amount) override
    {
        int notesNeed = amount / 1000;

        if (notesNeed > numnotes)
            notesNeed = numnotes;

        numnotes -= notesNeed;

        if (notesNeed > 0)
            cout << "Dispensing " << notesNeed << " notes of 1000\n";

        int remaining = amount - notesNeed * 1000;

        if (remaining > 0 && next)
            next->cashDispenser(remaining);
        else if (remaining > 0)
            cout << "Cannot dispense remaining amount: " << remaining << endl;
    }
};

class HundredHandler : public MoneyDispenser
{
    int numnotes;

public:
    HundredHandler(int notes) : numnotes(notes) {}

    void cashDispenser(int amount) override
    {
        int notesNeed = amount / 100;

        if (notesNeed > numnotes)
            notesNeed = numnotes;

        numnotes -= notesNeed;

        if (notesNeed > 0)
            cout << "Dispensing " << notesNeed << " notes of 100\n";

        int remaining = amount - notesNeed * 100;

        if (remaining > 0 && next)
            next->cashDispenser(remaining);
        else if (remaining > 0)
            cout << "Cannot dispense remaining amount: " << remaining << endl;
    }
};

class FiftyHandler : public MoneyDispenser
{
    int numnotes;

public:
    FiftyHandler(int notes) : numnotes(notes) {}

    void cashDispenser(int amount) override
    {
        int notesNeed = amount / 50;

        if (notesNeed > numnotes)
            notesNeed = numnotes;

        numnotes -= notesNeed;

        if (notesNeed > 0)
            cout << "Dispensing " << notesNeed << " notes of 50\n";

        int remaining = amount - notesNeed * 50;

        if (remaining > 0)
            cout << "Cannot dispense remaining amount: " << remaining << endl;
    }
};

int main()
{
    MoneyDispenser *M1 = new ThousandHandler(5);
    MoneyDispenser *M2 = new HundredHandler(15);
    MoneyDispenser *M3 = new FiftyHandler(20);

    M1->setHandler(M2);
    M2->setHandler(M3);

    M1->cashDispenser(2500);

    delete M1;
    delete M2;
    delete M3;
}
